### ISSUE TITLE
RangeError: Maximum call stack size exceeded (recursive path-finding script)

### ISSUE THREAD

**Issue by marcelbpunkt** [danielyxie/bitburner#4252]
When I run the below script (NS2) to find the path to a target node, I get this error:
```
RUNTIME ERROR
getPath.js@home (PID - 114)
Args: ["I.I.I.I"]

Maximum call stack size exceeded
stack:
RangeError: Maximum call stack size exceeded
at file:///home/marcel/.local/share/Steam/steamapps/common/Bitburner/resources/app/dist/main.bundle.js:19:59393
at Object.value [as scan] (file:///home/marcel/.local/share/Steam/steamapps/common/Bitburner/resources/app/dist/main.bundle.js:21:912755)
at getPathToServer (home/getPath.js:21:20)
at getPathToServer (home/getPath.js:31:22)
at getPathToServer (home/getPath.js:31:22)
at getPathToServer (home/getPath.js:31:22)
at getPathToServer (home/getPath.js:31:22)
at getPathToServer (home/getPath.js:31:22)
at getPathToServer (home/getPath.js:31:22)
at getPathToServer (home/getPath.js:31:22)
```
This is the script:
```javascript
/** @param {NS} ns */
export async function main(ns) {
	var target = ns.args[0];
	var rootNode = ns.args.length > 1 ? ns.args[1] : "home";
	var path = getPathToServer(ns, target, rootNode);

	ns.tprint(path.length ? path : `No path found from ${rootNode} to ${target}.`);
}

export function getPathToServer(ns, target, rootNode = "home") {
	var adjNodes = ns.scan(rootNode);

	// dead end
	if (!adjNodes.length) return [];

	// target is adjacent to this node
	if (adjNodes.includes(target)) return [target];

	// target is not adjacent to this node -> next recursion step
	for (var node of adjNodes) {
		var pathFromHere = getPathToServer(ns, target, node);
		if (pathFromHere.length) {
			// path to target found
			return [node].concat(pathFromHere);
		}
		// else do nothing and continue to next node since we hit a dead end
	}

	// there's no path from here
	return [];
}
```
It's a recursive Depth-First Search so I do know *why* this runtime error appears. I'd just like to know if the call stack size is intentionally so low (the recursion depth is really not that big) and, if so, can I increase it somehow?

Game version: v2.1.0 (8f4636cb)
OS: Linux Mint 20.3 (Una)
Steam startup option: "Play --no-sandbox (diff flags for different linux d..."

Steps to reproduce (assuming the in-game script file name is "getPath.js"):
* Enter `run getPath.js I.I.I.I` on the home computer in sector-12 (should work with other nodes than `I.I.I.I` too depending on where they are and what paths the scripts takes on its way there).
* Runtime error appears. If it doesn't, try other target nodes.

**Comment by smeehrrr** [comment]
I don't see anything in here that's preventing backtracking, so I think you have infinite recursion?

**Comment by Tanimodori** [comment]
It seems you may forget to mark visited nodes so the DFS algorithm may stuck.

Print out the node will help debugging.

**Comment by marcelbpunkt** [comment]
I will bury myself in shame. Somehow in my head it seemed valid that I didn't need to do backtracking if I'm dealing with a tree. But first, I don't even know *for sure* that I'm dealing with a tree here, and second, *of course* `scan` will also return the node I just came from. Thanks for pointing out the obviously not obvious-to-me mistake ;)
I corrected the script and everything runs smoothly now so the ticket can be closed.
